tests/extmod/machine_uart_irq_txidle.py

```diff
@@ -0,0 +1,69 @@
+# Test machine.UART.IRQ_TXIDLE firing after a transmission.
+# Does not require any external connections.
+
+try:
+    from machine import UART
+
+    UART.IRQ_TXIDLE
+except (ImportError, AttributeError):
+    print("SKIP")
+    raise SystemExit
+
+import time, sys
+
+# Configure pins based on the target.
+if "rp2" in sys.platform:
+    uart_id = 0
+    tx_pin = "GPIO0"
+    rx_pin = "GPIO1"
+    min_window = 1
+elif "samd" in sys.platform and "ItsyBitsy M0" in sys.implementation._machine:
+    uart_id = 0
+    tx_pin = "D1"
+    rx_pin = "D0"
+    # For SAMD delay_ms has to be used for the trailing window, and the
+    # mininmal time is 11 ms to allow for scheduling.
+    min_window = 11
+elif "samd" in sys.platform and "ItsyBitsy M4" in sys.implementation._machine:
+    uart_id = 3
+    tx_pin = "D1"
+    rx_pin = "D0"
+    min_window = 11
+elif "mimxrt" in sys.platform and "Teensy 4" in sys.implementation._machine:
+    uart_id = 1
+    tx_pin = None
+    min_window = 0
+elif "nrf" in sys.platform:
+    uart_id = 0
+    tx_pin = None
+    min_window = 0
+else:
+    print("SKIP")
+    raise SystemExit
+
+
+def irq(u):
+    print("IRQ_TXIDLE:", bool(u.irq().flags() & u.IRQ_TXIDLE))
+
+
+text = "Hello World" * 20
+
+# Test that the IRQ is called after the write has completed.
+for bits_per_s in (2400, 9600, 115200):
+    if tx_pin is None:
+        uart = UART(uart_id, bits_per_s)
+    else:
+        uart = UART(uart_id, bits_per_s, tx=tx_pin, rx=rx_pin)
+
+    uart.irq(irq, uart.IRQ_TXIDLE)
+
+    bits_per_char = 10  # 1(startbit) + 8(bits) + 1(stopbit) + 0(parity)
+    start_time_us = (len(text) - 10) * bits_per_char * 1_000_000 // bits_per_s
+    window_ms = max(min_window, 20 * bits_per_char * 1_000 // bits_per_s + 1)
```

Comment:
It's probably worth commenting on the algorithm used in these two calculations.  Why `-10` for `start_time_us`, and why `20 *` for `window_ms`, etc?

Reply:
Added. The value of 20 could probably be smaller.  SAMD devices use busy waiting for sleep_ms(n)
for n <= 10 and for sleep_us(), preventing the IRQ to get scheduled. For these I added the minimum wait time. But I prepared a change for SAMD to make that obsolete. Will be in a separate PR.